test/integration/garbagecollector/garbage_collector_test.go

```diff
@@ -1276,3 +1288,91 @@ func testCRDDeletion(t *testing.T, ctx *testContext, ns *v1.Namespace, definitio
 		t.Fatalf("failed waiting for dependent %q (owned by %q) to be deleted", dependent.GetName(), owner.GetName())
 	}
 }
+
+func TestDiscoveryFailedOnGCStart(t *testing.T) {
+	ctx := setup(t, 0)
+	defer ctx.tearDown()
+	apiRegClient, clientSet := ctx.apiRegistrationClient, ctx.clientSet
+
+	// Create a fake apiservice to simulate discovery failed
+	apiservice, err := apiRegClient.APIServices().Create(context.TODO(), &apiregistration.APIService{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "APIService",
+			APIVersion: "apiregistration.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "v1.test.io",
+		},
+		Spec: apiregistration.APIServiceSpec{
+			Service: &apiregistration.ServiceReference{
+				Namespace: "default",
+				Name:      "test",
+			},
+			Group:                "test.io",
+			Version:              "v1",
+			GroupPriorityMinimum: 1,
+			VersionPriority:      1,
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create apiservice: %v", err)
+	}
+	defer apiRegClient.APIServices().Delete(context.TODO(), apiservice.Name, metav1.DeleteOptions{})
+
+	ns := createNamespaceOrDie("gc-discovery-fail", clientSet, t)
+	defer deleteNamespaceOrDie(ns.Name, clientSet, t)
+
+	testBlockingOwnerRefDoesBlock(t, ctx, ns, 1*time.Second)
+}
+
+func TestCacheSyncFailedOnGCStart(t *testing.T) {
+	ctx := setup(t, 0)
+	defer ctx.tearDown()
+	apiExtensionClient, dynamicClient, clientSet := ctx.apiExtensionClient, ctx.dynamicClient, ctx.clientSet
+
+	ns := createNamespaceOrDie("gc-cache-sync-fail", clientSet, t)
+	defer deleteNamespaceOrDie(ns.Name, clientSet, t)
+
+	// Create a def with `v1beta1` version and a cr of the def, then update the def added
+	// with "v1" version as default and a fake conversion webhook, it will cause the
+	// cache syncing failed.
+	def, dc := createRandomCustomResourceDefinition(t, apiExtensionClient, dynamicClient, ns.Name)
+	_, err := dc.Create(context.TODO(), newCRDInstance(def, ns.Name, names.SimpleNameGenerator.GenerateName("foo-")), metav1.CreateOptions{})
```

Comment:
can we create the CRD with v1beta1 (served=true,storage=true) and v1(served=true,storage=false) initially rather than updating the CRD? the write to the v1beta1 endpoint shouldn't exercise or be blocked by conversion, but GC will prefer v1 so will hit the list errors

Reply:
Emm, it should also work.

Reply:
@liggitt, it can't work because the webhook will be called when creating the CRD.

Reply:
the conversion webhook shouldn't be called for create requests using the same API as the storage version